crates/worker/src/partition/mod.rs

```diff
@@ -517,6 +560,224 @@ where
+                    )
+                    .await
+                {
+                    Ok(()) => Ok(PartitionProcessorRpcResponse::Appended),
+                    Err(Error::NotLeader) => {
+                        Err(PartitionProcessorRpcError::NotLeader(self.partition_id))
+                    }
+                    Err(e) => Err(PartitionProcessorRpcError::Internal(e.to_string())),
+                };
+
+                self.respond_to_rpc(response_tx.prepare(rpc_response));
+            }
+        };
+    }
+
+    fn respond_to_rpc(
+        &self,
+        outgoing: Outgoing<
+            Result<PartitionProcessorRpcResponse, PartitionProcessorRpcError>,
+            HasConnection,
+        >,
+    ) {
+        // ignore shutdown errors
+        let _ = self.task_center.spawn(
+            TaskKind::Disposable,
+            "partition-processor-rpc-response",
+            None,
+            async move { outgoing.send().await.map_err(Into::into) },
+        );
+    }
+
+    async fn handle_rpc_get_invocation_output(
+        &self,
+        request_id: PartitionProcessorRpcRequestId,
+        invocation_query: InvocationQuery,
+        partition_store: &mut PartitionStore,
+    ) -> Result<PartitionProcessorRpcResponse, StorageError> {
+        // We can handle this immediately by querying the partition store, no need to go through proposals
+        // TODO update this logic when finishing up https://github.com/restatedev/restate/pull/2035
+        let invocation_id = match invocation_query {
+            InvocationQuery::Invocation(iid) => iid,
+            InvocationQuery::Workflow(sid) => {
+                match partition_store.get_virtual_object_status(&sid).await? {
+                    VirtualObjectStatus::Locked(iid) => iid,
+                    VirtualObjectStatus::Unlocked => {
+                        return Ok(PartitionProcessorRpcResponse::NotFound)
+                    }
+                }
+            }
+            InvocationQuery::IdempotencyId(iid) => {
+                match partition_store.get_idempotency_metadata(&iid).await? {
+                    Some(idempotency_metadata) => idempotency_metadata.invocation_id,
+                    None => return Ok(PartitionProcessorRpcResponse::NotFound),
+                }
+            }
+        };
+
+        let invocation_status = partition_store
+            .get_invocation_status(&invocation_id)
+            .await?;
+
+        match invocation_status {
+            InvocationStatus::Free => Ok(PartitionProcessorRpcResponse::NotFound),
+            is if is.idempotency_key().is_none()
+                && is
+                    .invocation_target()
+                    .map(InvocationTarget::invocation_target_ty)
+                    != Some(InvocationTargetType::Workflow(
+                        WorkflowHandlerType::Workflow,
+                    )) =>
+            {
+                Ok(PartitionProcessorRpcResponse::NotSupported)
```

Comment:
Could we also return `NotReady` if the status is neither free nor completed independent of this special casing?

Reply:
I would like to retain this special casing, as back in the handler we use this special case to give back the user a nicer descriptive error. Look at `HandlerError::UnsupportedGetOutput`.